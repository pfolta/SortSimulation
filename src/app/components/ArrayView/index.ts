export type { ArrayViewProps } from "@/app/components/ArrayView/ArrayView";
export { default as ArrayView } from "@/app/components/ArrayView/ArrayView";

export type { ImageArrayViewProps } from "@/app/components/ArrayView/ImageArrayView";
export { default as ImageArrayView } from "@/app/components/ArrayView/ImageArrayView";

export { default as NumberArrayView } from "@/app/components/ArrayView/NumberArrayView";
