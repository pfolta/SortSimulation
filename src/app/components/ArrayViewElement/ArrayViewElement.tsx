import styled from "styled-components";

interface ArrayViewElementProps {
    value: number;
}

const ArrayViewElement = styled.div.withConfig({
    shouldForwardProp: (prop) => ["children", "style"].includes(prop)
})<ArrayViewElementProps>`
    background-color: #4c566a;
    color: white;

    transition: transform 100ms ease-out;

    :hover {
        transform: translateY(-1rem);

        /* Prevent jitter on hover, see https://codepen.io/csilverman/post/fixing-the-jitter-bug */
        ::before {
            content: "";

            position: absolute;
            top: 0;
            left: 0;
            right: 0;
            bottom: -1rem;

            z-index: -1;
        }
    }
`;

export default ArrayViewElement;
export { ArrayViewElementProps };
