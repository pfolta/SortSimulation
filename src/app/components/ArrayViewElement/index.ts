export type { ArrayViewElementProps } from "@/app/components/ArrayViewElement/ArrayViewElement";
export { default as ArrayViewElement } from "@/app/components/ArrayViewElement/ArrayViewElement";

export type { ImageElementProps } from "@/app/components/ArrayViewElement/ImageElement";
export { default as ImageElement, StyledImageElement } from "@/app/components/ArrayViewElement/ImageElement";

export type { NumberElementProps } from "@/app/components/ArrayViewElement/NumberElement";
export { default as NumberElement, StyledNumberElement } from "@/app/components/ArrayViewElement/NumberElement";
