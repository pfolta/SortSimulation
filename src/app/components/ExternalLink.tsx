import styled from "styled-components";

const ExternalLink = styled.a.attrs({
    target: "_blank",
    rel: "noreferrer"
})`
    :hover {
        text-decoration: none;
    }
`;

export default ExternalLink;
