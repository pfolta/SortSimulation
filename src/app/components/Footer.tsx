import React from "react";
import styled from "styled-components";

const StyledFooter = styled.footer`
    background-color: #e5e9f0;
    color: #4c566a;

    font-size: 0.875rem;

    padding: 1rem;
`;

const Footer = (): JSX.Element => <StyledFooter>Â© Peter Folta. All rights reserved.</StyledFooter>;

export default Footer;
