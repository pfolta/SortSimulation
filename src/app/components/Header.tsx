import React from "react";
import { HelpCircle, Moon, Settings, Sun } from "react-feather";
import styled from "styled-components";

import { useDispatch, useSelector } from "@/app/hooks";
import { toggleTheme } from "@/app/store";
import { themes } from "@/app/theme";

const StyledHeader = styled.header`
    background-color: peachpuff;

    display: flex;
    align-items: center;
    justify-content: space-between;

    position: sticky;
    top: 0;
`;

const Controls = styled.ul`
    display: flex;
`;

const Control = styled.li`
    list-style-type: none;
`;

const Button = styled.button`
    background: none;
    color: #2e3440;

    border: none;
    border-radius: 0.25rem;

    display: flex;
    padding: 0.25rem;

    :hover {
        background-color: rgba(0, 0, 0, 0.25);
    }
`;

const Header = (): JSX.Element => {
    const { theme } = useSelector((state) => state.settings);
    const dispatch = useDispatch();

    return (
        <StyledHeader>
            <h1>SortSimulation</h1>
            <Controls>
                <Control>
                    <Button onClick={() => dispatch(toggleTheme())}>{themes[theme] === themes.dark ? <Sun /> : <Moon />}</Button>
                </Control>
                <Control>
                    <Button>
                        <Settings />
                    </Button>
                </Control>
                <Control>
                    <Button>
                        <HelpCircle />
                    </Button>
                </Control>
            </Controls>
        </StyledHeader>
    );
};

export default Header;
