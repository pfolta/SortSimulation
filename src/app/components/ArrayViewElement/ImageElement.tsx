import React from "react";
import styled from "styled-components";

import { ArrayViewElement, ArrayViewElementProps } from "@/app/components/ArrayViewElement";

interface ImageElementProps extends ArrayViewElementProps {
    image: string;
}

const StyledImageElement = styled(ArrayViewElement)<ImageElementProps>`
    background-image: url("${({ image }) => image}");
    background-position: ${({ array, index }) => ((array[index] - 1) / (array.length - 1)) * 100}% 0%;
    background-repeat: no-repeat;
    background-size: ${({ array }) => array.length * 100}% 100%;

    height: 100%;
`;

const ImageElement = (props: ImageElementProps): JSX.Element => <StyledImageElement {...props} />;

export default ImageElement;
export { ImageElementProps, StyledImageElement };
