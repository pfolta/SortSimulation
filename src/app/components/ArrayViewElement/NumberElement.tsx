import React from "react";
import Measure from "react-measure";
import styled from "styled-components";

import { ArrayViewElement, ArrayViewElementProps } from "@/app/components/ArrayViewElement";

interface NumberElementProps extends ArrayViewElementProps {
    maxValue: number;
}

interface StyledNumberElementProps extends NumberElementProps {
    contentWidth?: number;
}

const StyledNumberElement = styled(ArrayViewElement).attrs<StyledNumberElementProps>(({ array, index, maxValue, contentWidth }) => ({
    style: {
        fontSize: computeFontSize(contentWidth) + "px",
        height: (array[index] / maxValue) * 100 + "%"
    }
}))<StyledNumberElementProps>`
    font-size: 0;
    text-align: center;
`;

const StyledNumberElementLabel = styled.span`
    display: block;
    height: 100%;
    overflow: hidden;
`;

const computeFontSize = (width?: number): number => {
    const size = width ? width / 2 : 0;
    return size < 8 ? 0 : size;
};

const NumberElement = (props: NumberElementProps): JSX.Element => (
    <Measure>
        {({ contentRect, measureRef }) => (
            <StyledNumberElement ref={measureRef} {...props} contentWidth={contentRect.entry?.width}>
                <StyledNumberElementLabel>{props.array[props.index]}</StyledNumberElementLabel>
            </StyledNumberElement>
        )}
    </Measure>
);

export default NumberElement;
export { NumberElementProps, StyledNumberElement, StyledNumberElementLabel };
