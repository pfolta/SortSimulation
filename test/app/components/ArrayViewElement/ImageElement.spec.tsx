import React from "react";
import { create } from "react-test-renderer";

import { ImageElement } from "@/app/components/ArrayViewElement";

describe("ImageElement", () => {
    const array = [4, 2, 5, 1, 3];
    const image = "/path/to/image.jpg";

    it("renders correctly", () => {
        array
            .map((value) => create(<ImageElement value={value} size={array.length} image={image} />))
            .forEach((element) => expect(element.toJSON()).toMatchSnapshot());
    });

    it("uses the specified image as a background image", () => {
        array
            .map((value) => create(<ImageElement value={value} size={array.length} image={image} />))
            .forEach((element) => expect(element.toJSON()).toHaveStyleRule("background-image", `url("${image}")`));
    });

    it("sizes the image according to the number of elements in the array", () => {
        array
            .map((value) => create(<ImageElement value={value} size={array.length} image={image} />))
            .forEach((element) => expect(element.toJSON()).toHaveStyleRule("background-size", `${array.length * 100}% 100%`));
    });
});
