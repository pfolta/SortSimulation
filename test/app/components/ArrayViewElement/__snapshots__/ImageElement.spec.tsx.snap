// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageElement renders correctly 1`] = `
.c0 {
  background-color: #4c566a;
  color: white;
  -webkit-transition: -webkit-transform 100ms ease-out;
  -webkit-transition: transform 100ms ease-out;
  transition: transform 100ms ease-out;
}

.c0:hover {
  -webkit-transform: translateY(-1rem);
  -ms-transform: translateY(-1rem);
  transform: translateY(-1rem);
}

.c0:hover::before {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: -1rem;
  z-index: -1;
}

.c1 {
  background-image: url("/path/to/image.jpg");
  background-repeat: no-repeat;
  background-size: 500% 100%;
  height: 100%;
}

<div
  className="c0 c1"
  style={
    Object {
      "backgroundPosition": "75% 0%",
    }
  }
/>
`;

exports[`ImageElement renders correctly 2`] = `
.c0 {
  background-color: #4c566a;
  color: white;
  -webkit-transition: -webkit-transform 100ms ease-out;
  -webkit-transition: transform 100ms ease-out;
  transition: transform 100ms ease-out;
}

.c0:hover {
  -webkit-transform: translateY(-1rem);
  -ms-transform: translateY(-1rem);
  transform: translateY(-1rem);
}

.c0:hover::before {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: -1rem;
  z-index: -1;
}

.c1 {
  background-image: url("/path/to/image.jpg");
  background-repeat: no-repeat;
  background-size: 500% 100%;
  height: 100%;
}

<div
  className="c0 c1"
  style={
    Object {
      "backgroundPosition": "25% 0%",
    }
  }
/>
`;

exports[`ImageElement renders correctly 3`] = `
.c0 {
  background-color: #4c566a;
  color: white;
  -webkit-transition: -webkit-transform 100ms ease-out;
  -webkit-transition: transform 100ms ease-out;
  transition: transform 100ms ease-out;
}

.c0:hover {
  -webkit-transform: translateY(-1rem);
  -ms-transform: translateY(-1rem);
  transform: translateY(-1rem);
}

.c0:hover::before {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: -1rem;
  z-index: -1;
}

.c1 {
  background-image: url("/path/to/image.jpg");
  background-repeat: no-repeat;
  background-size: 500% 100%;
  height: 100%;
}

<div
  className="c0 c1"
  style={
    Object {
      "backgroundPosition": "100% 0%",
    }
  }
/>
`;

exports[`ImageElement renders correctly 4`] = `
.c0 {
  background-color: #4c566a;
  color: white;
  -webkit-transition: -webkit-transform 100ms ease-out;
  -webkit-transition: transform 100ms ease-out;
  transition: transform 100ms ease-out;
}

.c0:hover {
  -webkit-transform: translateY(-1rem);
  -ms-transform: translateY(-1rem);
  transform: translateY(-1rem);
}

.c0:hover::before {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: -1rem;
  z-index: -1;
}

.c1 {
  background-image: url("/path/to/image.jpg");
  background-repeat: no-repeat;
  background-size: 500% 100%;
  height: 100%;
}

<div
  className="c0 c1"
  style={
    Object {
      "backgroundPosition": "0% 0%",
    }
  }
/>
`;

exports[`ImageElement renders correctly 5`] = `
.c0 {
  background-color: #4c566a;
  color: white;
  -webkit-transition: -webkit-transform 100ms ease-out;
  -webkit-transition: transform 100ms ease-out;
  transition: transform 100ms ease-out;
}

.c0:hover {
  -webkit-transform: translateY(-1rem);
  -ms-transform: translateY(-1rem);
  transform: translateY(-1rem);
}

.c0:hover::before {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: -1rem;
  z-index: -1;
}

.c1 {
  background-image: url("/path/to/image.jpg");
  background-repeat: no-repeat;
  background-size: 500% 100%;
  height: 100%;
}

<div
  className="c0 c1"
  style={
    Object {
      "backgroundPosition": "50% 0%",
    }
  }
/>
`;
